src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,288 @@
+  out->mutable_mfa_state()->CopyFrom(hypers_);
+  return out;
+}
+
+void MFAHierarchy::sample_full_cond(bool update_params) {
+  assert(update_params != true);  // should never be true
+  if (this->card == 0) {
+    // No posterior update possible
+    sample_prior();
+  } else {
+    sample_Eta();
+    sample_mu();
+    sample_psi();
+    sample_Lambda();
+  }
+}
+
+void MFAHierarchy::sample_Eta() {
+  auto& rng = bayesmix::Rng::Instance().get();
+
+  Eigen::MatrixXd Sigmaeta(
+      (Eigen::MatrixXd::Identity(hypers->q, hypers->q) +
+       state.Lambda.transpose() *
+           Eigen::MatrixXd(state.psi.cwiseInverse().asDiagonal()) *
+           state.Lambda)
+          .inverse());
+
+  Eigen::MatrixXd temp(state.psi.cwiseInverse().asDiagonal());
+  /*
+  std::cout << "Sigmaeta: " << Sigmaeta.rows() << ", " << Sigmaeta.cols() << std::endl;
+  std::cout << "temp: " << temp.rows() << ", " << temp.cols() << std::endl;
+  std::cout << "lambda: " << state.Lambda.rows() << ", " << state.Lambda.cols() << std::endl;
+  Eigen::VectorXd difference = data[0] - state.mu;
+  std::cout << "difference: " << difference.size() << std::endl;
+
+  std::cout << stan::math::multi_normal_rng(
+        Sigmaeta * (state.Lambda.transpose()) * 
+            temp * 
+            difference, Sigmaeta, rng) << std::endl;
+
+  std::cout << "Eta: " << state.Eta.rows() << ", " << state.Eta.cols() << std::endl;
+  */
+  // DA SISTEMARE INIZIALLIZZANDO OGNI VOLTA CORRETTAMENTE GLI ELEMENTI MANCANTI
+  if (state.Eta.rows() != card) {
+    state.Eta.resize(card, state.Eta.cols());
+    // PER ORA INIZIALIZZO COSI'
+    /*
+    for (size_t i = 0; i < card; i++) {
+      for (size_t j = 0; j < state.Eta.cols(); j++) {
+        state.Eta(i, j) = stan::math::normal_rng(0, 1, rng);
+      }
+    }
+    */
+    state.Eta = Eigen::MatrixXd::Zero(card, state.Eta.cols());
+  }
+  for (size_t i = 0; i < card; i++) {
+    state.Eta.row(i) = (stan::math::multi_normal_rng(
+        Sigmaeta * (state.Lambda.transpose()) * 
+            temp * 
+            (data[i] - state.mu),
+        Sigmaeta, rng));
+  }
+}
+
+void MFAHierarchy::sample_mu() {
+  auto& rng = bayesmix::Rng::Instance().get();
+
+  Eigen::MatrixXd Sigmamu =
+      (hypers->phi * Eigen::MatrixXd::Identity(p, p) +
+
+       card *
+           Eigen::MatrixXd(state.psi.cwiseInverse().asDiagonal())).inverse();
+
+  Eigen::VectorXd Somma = Eigen::VectorXd::Zero(p);
+
+  for (size_t i = 0; i < card; i++) {
+    Eigen::VectorXd riga = state.Eta.row(i);
+    Somma += state.Lambda * riga;
+  }
+  //std::cout << "Somma dei dati"<< Somma << "Fine"<< std::endl;
+  Eigen::VectorXd mumean =
+      Sigmamu * (hypers->phi * hypers->mutilde +
+                 Eigen::MatrixXd(state.psi.cwiseInverse().asDiagonal()) *
+                     (data_sum - Somma));
+
+
+  state.mu = stan::math::multi_normal_rng(mumean, Sigmamu, rng);
+}
+
+void MFAHierarchy::sample_Lambda() {
+  auto& rng = bayesmix::Rng::Instance().get();
```

Comment:
Data is now stored in a std::vector of Eigen::VectorXd, but in sample_Lambda(), we need to access both to rows and columns of the data matrix, so this is probably not ideal, for now at each iteration we create this temporary matrix, but this can't be a good solution. Should we just store data in an Eigen::Matrix? Or Adding and removing data would be too costly? We have not yet come up with an optimal solution

Reply:
There is no "one-rule-fits-all" answer. The only way is to benchmark both alternatives, varying the dimension and number of data. There will be probably some situations in which the first one is faster and vice-versa. Once you have a nice benchmark we can decide for the alternative that is more likely to help in real-case scenarios

Reply:
Why should a non-matrix solution ever be more efficient than the alternative? We might initialize a zero matrix with the largest possible dimensions, e.g. the number of data points, then access appropriate rows and columns using class member indexes, such as the already existing `card`, so that we don't access rows without actual data. But yeah, we will see